x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java

```diff
@@ -236,4 +244,35 @@ public Verifier(Metrics metrics) {
 
         return failures;
     }
-}
\ No newline at end of file
+
+    private void checkJoinKeyTypes(LogicalPlan plan, Set<Failure> localFailures) {
+        if (plan instanceof Join) {
+            Join join = (Join) plan;
+            List<KeyedFilter> queries = join.queries();
+            KeyedFilter until = join.until();
+            // pick first query and iterate its keys
+            KeyedFilter first = queries.get(0);
+            List<? extends NamedExpression> keys = first.keys();
+            for (int keyIndex = 0; keyIndex < keys.size(); keyIndex++) {
+                NamedExpression currentKey = keys.get(keyIndex);
+                for (int i = 1; i < queries.size(); i++) {
+                    KeyedFilter filter = queries.get(i);
+                    doCheckKeyTypes(join, localFailures, currentKey, filter.keys().get(keyIndex));
+                    if (until.keys().isEmpty() == false) {
+                        doCheckKeyTypes(join, localFailures, currentKey, until.keys().get(keyIndex));
+                    }
+                }
+            }
+        }
+    }
+
+    private static void doCheckKeyTypes(Join join, Set<Failure> localFailures, NamedExpression expectedKey, NamedExpression currentKey) {
+        if (DataTypes.areCompatible(expectedKey.dataType(), currentKey.dataType()) == false) {
+            localFailures.add(fail(currentKey, "{} key [{}] type [{}] is incompatible with key [{}] type [{}]",
```

Comment:
I would have chosen a slightly different error message, something that would made clearer that the incompatible types are to blame: `key [X] type [Y] has an incompatible type with key [A] type [B]`.

Reply:
What's unclear about the current message?
`Sequence key[X] type[Y] is incompatible with key [A] type [B]` vs 
`key[X] type[Y] has an incompatible type with key [A] type [B]`.

The messages are the same length however I find the former more informative.